### PR TITLE
8258192: Obsolete the CriticalJNINatives flag

### PR DESCRIPTION
**Pull request by coleenp**
This change removes the disabled CriticalJNINatives code, and the flag now gives an obsolete message.
Tested with tier1 on cpus x64, aarch64, and builds on linux-x86-open,linux-s390x-open,linux-arm32-debug,linux-ppc64le-debug.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8258192](https://bugs.openjdk.java.net/browse/JDK-8258192): Obsolete the CriticalJNINatives flag

### Reviewers
 * [Martin Doerr](https://openjdk.java.net/census#mdoerr) (@TheRealMDoerr - **Reviewer**) ⚠️ Review applies to 9b8ff9ae1752de0fb0134de107de07f1a8955b37
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to 9b8ff9ae1752de0fb0134de107de07f1a8955b37

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6343/head:pull/6343` \
`$ git checkout pull/6343`

Update a local copy of the PR: \
`$ git checkout pull/6343` \
`$ git pull https://git.openjdk.java.net/jdk pull/6343/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6343`

View PR using the GUI difftool: \
`$ git pr show -t 6343`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6343.diff">https://git.openjdk.java.net/jdk/pull/6343.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp [hunk 1]
```diff
@@ -1823,7 +1743,7 @@ nmethod *SharedRuntime::generate_native_wrapper(MacroAssembler *masm,
   //
   // NW     [ABI_REG_ARGS]             <-- 1) R1_SP
   //        [outgoing arguments]       <-- 2) R1_SP + out_arg_slot_offset
-  //        [oopHandle area]           <-- 3) R1_SP + oop_handle_offset (save area for critical natives)
+  //        [oopHandle area]           <-- 3) R1_SP + oop_handle_offset (save area for critical natives) ?
```
Comment (shipilev) [inline src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp:1746]: `?`. The comment `(save area for critical natives)` must be redundant now.

File: src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp [hunk 2]
```diff
@@ -1592,55 +1546,23 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
...
-    }
-  } else {
-    for (int i = 0; i < total_in_args; i++) {
-      if (in_sig_bt[i] == T_ARRAY) {
-        total_c_args++;
-      }
-    }
+  int total_c_args = total_in_args+1;
+  if (method->is_static()) {
+    total_c_args++;
```
Comment (shipilev) [inline src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp:1551]: In this patch, sometimes we keep the if structure, like here, but in other places, we replace this with:
```
  int total_c_args = total_in_args + (method->is_static() ? 2 : 1)
```
Should probably stick with a single style.

File: src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp [hunk 3]
```diff
@@ -1886,44 +1774,23 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
...
   tmp_vmreg.set2(rbx->as_VMReg());
 
-  if (!is_critical_native) {
-    for (int i = total_in_args - 1, c_arg = total_c_args - 1; i >= 0; i--, c_arg--) {
-      arg_order.push(i);
-      arg_order.push(c_arg);
-    }
-  } else {
-    // Compute a valid move order, using tmp_vmreg to break any cycles
-    ComputeMoveOrder cmo(total_in_args, in_regs, total_c_args, out_regs, in_sig_bt, arg_order, tmp_vmreg);
```
Comment (shipilev) [inline src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp:1905]: `ComputeMoveOrder` is still used somewhere, or?

File: src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp [hunk 4]
```diff
@@ -1886,44 +1774,23 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
...
-      temploc = i;
-      continue;
-    } else if (i == -1) {
-      assert(is_critical_native, "should only be required for critical natives");
-      // Read from the temporary location
-      assert(temploc != -1, "must be valid");
-      i = temploc;
-      temploc = -1;
-    }
+    assert (c_arg != -1, "wrong direction");
```
Comment (shipilev) [inline src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp:1793]: `assert (c_arg != -1 && i != -1, "wrong direction");`?

File: src/hotspot/share/runtime/sharedRuntime.cpp [hunk 5]
```diff
@@ -3014,11 +3014,6 @@ void AdapterHandlerLibrary::create_native_wrapper(const methodHandle& method) {
   assert(method->is_method_handle_intrinsic() ||
          method->has_native_function(), "must have something valid to call!");
 
-  if (CriticalJNINatives && !method->is_method_handle_intrinsic()) {
-    // We perform the I/O with transition to native before acquiring AdapterHandlerLibrary_lock.
-    critical_entry = NativeLookup::lookup_critical_entry(method);
```
Comment (shipilev) [inline src/hotspot/share/runtime/sharedRuntime.cpp:3019]: `critical_entry` variable is now redundant?

**Review by TheRealMDoerr — CHANGES_REQUESTED** [review]
Thanks for taking care of all platforms. `move_ptr(MacroAssembler*, VMRegPair, VMRegPair, int)` needs to get removed to avoid build warnings on PPC64 and s390.

File: src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp:1746]: I didn't know if the save area is still needed for something else, which is why I left the ?.  I can remove the comment but haven't made any  substantial changes here.  I'm not sure if they're needed or not, but I can't test them if I made them.

File: src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp:1551]: Ok, that's a good suggestion.  Some platforms have a method_is_static boolean and some don't, so I didn't clean up the platforms that do that later in a different way (or inconsistently).

File: src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp (hunk 1 above)
Comment (shipilev) [reply src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp:1746]: I mean, you did the same here: https://github.com/openjdk/jdk/pull/6343/files#diff-060e534de775616a893aa969f3639e53666cda9e93bed7c3a3c14b9cdc4cdba0L1553-L1554 -- and that change is understandable.

File: src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp (hunk 3 above)
Comment (coleenp) [reply src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp:1905]: Yes, it's used in
cpu/x86/universalUpcallHandler_x86_64.cpp:  SharedRuntime::compute_move_order(in_sig_bt,

File: src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp (hunk 4 above)
Comment (coleenp) [reply src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp:1793]: removed.

File: src/hotspot/share/runtime/sharedRuntime.cpp (hunk 5 above)
Comment (coleenp) [reply src/hotspot/share/runtime/sharedRuntime.cpp:3019]: removed, thanks for spotting that.

**Review by coleenp — COMMENTED** [review]
Thanks for reviewing, Aleksey.  I made the changes and will retest.

**Comment by coleenp** [comment]
Thanks for finding move_ptr @TheRealMDoerr.

File: src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp:1746]: Looking further, the area is needed (stores oops there), but not the comment.

File: src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp (hunk 3 above)
Comment (JornVernee) [reply src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp:1905]: FWIW, I have a change in panama-foreign repo that replaces that use with a custom class. Will remove ComputeMoveOrder there as well, and it should be completely gone after the next JEP integration, probably in 19.

**Review by TheRealMDoerr — APPROVED** [review]
LGTM. Thanks!

File: src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp (hunk 3 above)
Comment (coleenp) [reply src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp:1905]: Ok, thanks Jorn.

**Comment by coleenp** [comment]
Thanks for reviewing @TheRealMDoerr .  @shipilev does this look good now?

**Review by shipilev — APPROVED** [review]
Yeah, I am fine with this.

**Comment by coleenp** [comment]
Thanks @shipilev .  All the GHA passed after resolving above merge conflict.
/integrate

**Pull request closed without merging**